Recursor: implement Negative Trust Anchor management and runtime Trust Anchor management

pieterlexis opened the pull request:
This implementation *might* be a bit blunt, but it works. Please review.

Closes #3902

rgacogne commented on pdns/rec_channel_rec.cc:
```diff
@@ -341,6 +342,62 @@ string doSetCarbonServer(T begin, T end)
+  if(begin == end)
+    return "No NTA specified, doing nothing\n";
+
+  auto luaconf = g_luaconfs.getLocal();
```
It looks like `luaconf` is unused.

pieterlexis replied:
a leftover of an earlier incantation, removed

rgacogne commented on pdns/validate.cc:
```diff
@@ -165,6 +165,19 @@ cspmap_t harvestCSPFromRecs(const vector<DNSRecord>& recs)
+        LOG("Found a Negative Trust Anchor for "<<negAnchor.first<<", which was added with the message '"<<negAnchor.second<<"'. Returning Insecure."<<endl);
+        return NTA;
+      }
+
```
Shouldn't we check whether there is a (positive) trust anchor below? If I understand correctly section 1.1 of rfc7646, we would be expected to resume validation starting with this trust anchor, right?

pieterlexis commented:
Fixed @rgacogne's comments and added a bugfix that ensures the use of the correct, configured trust anchor (we only used the root trust anchor before)

pieterlexis commented:
an addition on the bugfix regarding positive trust anchors: this makes the Recursor behave as specified in [RFC 6840 §C.1](https://tools.ietf.org/html/rfc6840#appendix-C.1). i.e. using the closest encounter when validating responses.

pieterlexis commented:
Updated to also add runtime trust anchor management

Habbie commented on docs/manpages/rec_control.1.md:
```diff
@@ -43,9 +43,23 @@ To dump the cache to disk, execute:
 
+clear-nta [*DOMAIN*]...
+:    Remove Negative Trust Anchor for one or more *DOMAIN*s. If *DOMAIN* is
+     empty, remove all configured Negative Trust Anchors.
```
'empty' or 'no domain specified'? I think we might want to be explicit here, use '*' or something. If somebody automates setting/clearing of NTAs it's pretty easy to accidentally pass an empty list

Habbie commented:
I approve of this PR!

Habbie commented:
(one nit)
